### ISSUE TITLE
Improve typing of custom req/resp type

### ISSUE THREAD

**Issue by vytas7** [falconry/falcon#2372]
### Discussed in https://github.com/falconry/falcon/discussions/2370

<div type='discussions-op-text'>

<sup>Originally posted by **jkmnt** October 14, 2024</sup>
Testing f4 beta with the our falcon-powered app. No runtime issues so far, but a few errors from the typechecker.

The `app.add_error_handler` and `app.add_error_handler` expects the callables with signature of  [falcon.Request, falcon.Response, ...etc]. Our app is using the custom Request class:
```python
falcon.App(request_type=MyRequest)

def my_error_handler(req: MyRequest, ...): ...

```

Maybe the request/response types should be typed as generics bound to the base type ?
Something like
```
TReq = TypeVar('TReq', bound=Request)

class App(Generic[TReq, TResp]):
    def __init__( self,  request_type: Optional[type[TReq]] = None, ...): ...

    def add_error_handler(self, exception, handler: Callable[[TReq, TResp, TCtx, ...], None]): ...
```

The same issue is with `before`/`after` hook adders . Yes, it's way harder to type since they know nothing about the app class.</div>

**Comment by CaselIT** [comment]
> BTW, there was some (missed by mypy) errors reported by pyright. I think I'll try to fix some and submit a separate PR.

that would be great. Feel free to also open an issue when you do that.
also if you know tox and github action we could add a gate with it too, so that it stays tested

**Comment by vytas7** [comment]
I have just created an issue to track the Pyright gate: #2373.

**Comment by jkmnt** [comment]
Another thing related to the typing:
I noticed the middleware type is just the `object`. It's the user-facing interface so imho typing should be better.

I suggest of something like
```python
class MwWithReqHandler(Protocol):
    def process_request(self, req: Request, resp: Response) -> None: ...

class MwWithRespHandler(Protocol):
    def process_response(self, req: Request, resp: Response, resource: object, req_succeeded: bool) -> None: ...

class MwWithProcessHandler(Protocol):
    def process_resource(self, req: Request, resp: Response, resource: object, params: dict[str, Any]) -> None: ...

Middleware = Union[MwWithReqHandler, MwWithRespHandler, MwWithProcessHandler]
```

This is not perfect but better than nothing. The bad thing is that it's enough to have one conforming (with others nonconformin) methods to pass the type check. The good thing: the user may inherit it's middleware class from these protocols (kind of abc but w/o runtime overhead) and get full typesafety.

```python
class MyMiddleware(MwWithReqHandler, MwWithRespHandler):
    def process_req(...)
    def process_resp(...)
```

**Comment by CaselIT** [comment]
I thought about it while doing that part, but if we type it as `Middleware = Union[MwWithReqHandler, MwWithRespHandler, MwWithProcessHandler]` won't that require that all 3 method are provided?

If not then that's for sure an option!

**Comment by jkmnt** [comment]
No afaik its enough to satisfy just one protocol of the union.

**Comment by CaselIT** [comment]
Seems like a good option then!

**Comment by CaselIT** [comment]
it probably makes sense to open a new issue with it. If you can work on a PR it would be great!

**Comment by vytas7** [comment]
Yes, let's divide and conquer these typing improvements. Like improvement of the middleware protocols could be one issue/PR :slightly_smiling_face:

**Comment by jkmnt** [comment]
> it probably makes sense to open a new issue with it. If you can work on a PR it would be great!

Ok, I'll draft the PR then.

**Comment by davetapley** [comment]
I managed to type this all up at https://github.com/falconry/falcon/issues/2397 before looking for this, but looks like I came to the same conclusions 🙈 

Thanks!

**Comment by vytas7** [comment]
It seems it is possible to do away without @jkmnt 's `SimpleApp` by overloading `App.__init__`, but it gets hairy pretty quickly, and it doesn't allow explicitly passing `request_type=None` (unless adding even more overloads? 🤯).

But maybe it could still be a good start?

**Comment by vytas7** [comment]
@jkmnt @CaselIT @davetapley This is my rough PoC which seems to work (not transplanted into Falcon yet):
```python
from typing import Generic, overload, TypeVar


class Request:
    def __init__(self, query: str) -> None:
        self._query = query

    @property
    def query(self) -> str:
        return self._query


class Response:
    result: str

    def __init__(self) -> None:
        self.result = ''


class FancyRequest(Request):
    @property
    def query(self) -> str:
        return f'!!! {self._query} !!!'


class FancyResponse(Response):
    """TODO: Maybe not so fancy after all?"""


TREQ = TypeVar('TREQ', bound=Request)
TRSP = TypeVar('TRSP', bound=Response)


class App(Generic[TREQ, TRSP]):
    name: str
    req_type: type[TREQ]
    resp_type: type[TRSP]

    @overload
    def __init__(self: 'App[Request, Response]', name: str) -> None: ...

    @overload
    def __init__(self: 'App[TREQ, Response]', name: str,
                 req_type: type[TREQ] | None = None) -> None: ...

    @overload
    def __init__(self: 'App[Request, TRSP]', name: str,
                 req_type: None = None,
                 resp_type: type[TRSP] | None = None) -> None: ...

    @overload
    def __init__(self, name: str,
                 req_type: type[TREQ] | None = None,
                 resp_type: type[TRSP] | None = None) -> None: ...

    def __init__(self, name: str,
                 req_type: type[TREQ] | None = None,
                 resp_type: type[TRSP] | None = None):
        self.name = name

        self.req_type = req_type or Request  # type: ignore[assignment]
        self.resp_type = resp_type or Response  # type: ignore[assignment]

    def handle(self, query: str) -> None:
        req = self.req_type(query)
        resp = self.resp_type()

        resp.result = f'{req.query} => Wow such result'

        print(resp.result)


if __name__ == '__main__':
    app = App('LLM')
    app.handle('Are you a robot?')

    app2 = App('FancyLLM', req_type=FancyRequest)
    app2.handle('Are you a robot?')

    app3 = App('FancyResp', resp_type=FancyResponse)
    app3.handle('Are you a robot?')

    app4 = App('FancyEverything',
               req_type=FancyRequest, resp_type=FancyResponse)
    app4.handle('Custom.')
```

Note that I still have no idea how to get rid of `# type: ignore[assigment]`, but it is well not a concern for the end user?

**Comment by CaselIT** [comment]
My only reserve is that I don't know how useful it may be this use case, since generally app is used only in one place of the code were defining routes.

Maybe we can try to solve just the original issue, meaning:
> The app.add_error_handler and app.add_error_handler expects the callables with signature of [falcon.Request, falcon.Response, ...etc]. Our app is using the custom Request class:

we could try doing something like the following (in 3.12+ syntax, but it should work also using typevars):
```py
    @overload
    def add_error_handler[RQ: Request, RE: Response, E: Exception](
        self,
        exception: Type[E],
        handler: Callable[[RQ, RE, E, Dict[str, Any]], None],
    ) -> None: ...
```

once we do that this example passes:
```py
from typing import Any
import falcon


class MR(falcon.Request):
    foo = 1


app = falcon.App(request_type=MR)


def handler(
    req: MR, resp: falcon.Response, exception: KeyError, params: dict[str, Any]
)-> None:
    print(req.foo)
    return


app.add_error_handler(KeyError, handler)

```

**Comment by vytas7** [comment]
@CaselIT if that's all what's needed, that's great.
Does it pass with `mypy --strict` though?

I couldn't get around the problem for `Protocol`s either, you can use generics, but it wouldn't know what to template on.

**Comment by vytas7** [comment]
I did some more prototyping with generics, and things seem to work on the conceptual level, albeit it is ugly as hell...

Another problem is that our ASGI `Request`, `Response`, `App` types violate (by design...) the [Liskov substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle) when subclassing the WSGI counterparts, for instance, `asgi.App.__call__()` becomes a coroutine function with different parameter signatures. But this can hopefully be addressed with `# type: ignore`s (already addressed by @CaselIT in the codebase, but I mean hopefully it doesn't pose new challenges when generalizing the App classes).

This approach, even if we succeed with it, will not make it into the Falcon 4.1 release, though.
